### PR TITLE
daemon: read-copy-update the daemon config

### PR DESCRIPTION
**Pull request by corhere**
**- What I did**
Made daemon config reloading transactional and almost entirely atomic. Config reloading will either succeed or leave the daemon state untouched. Reloading no longer pauses the world. Daemon operations see a consistent and immutable view of the daemon config, even when racing a reload.

**- How I did it**
- Made the `config.Config` struct copyable by removing the embedded mutexes
- Implemented read-copy-update of the daemon config to replace the embedded mutex for synchronization
  - Reading from daemon config is request-scoped, like contexts: any operations which need to access the daemon config capture a reference to an immutable snapshot and pass it into any callees
- Implemented two-phase commit to manage the side effects of config reloading
  - Side effects such as `registryService.ReplaceConfig()` are not necessarily atomic w.r.t. the reload transaction as they can take effect before the updated config is applied to `configStore`
- Modified how runtimes wrapper scripts are written so that existing wrapper scripts are no longer deleted on reload
  - reloads do not invalidate old versions of the config that concurrent daemon operations might be holding references to, which might reference old wrapper scripts
  - FS errors while updating the wrapper scripts can no longer leave containers or the daemon in a bad state

**- How to verify it**
CI

**- Description for the changelog**

- Configuration reloading has been made more robust: in the event that an error is encountered during the reload process, there is no longer a chance for some of the options to be applied anyway.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/config/registry.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"path/filepath"
+
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/docker/docker/registry"
+	"github.com/moby/buildkit/util/resolver"
+	resolverconfig "github.com/moby/buildkit/util/resolver/config"
+)
+
+// RegistryHosts returns registry configuration in containerd resolvers format.
+func (conf *Config) RegistryHosts() docker.RegistryHosts {
```
Comment (tianon) [inline daemon/config/registry.go:15]: Reviewer's note: this is just moved from `daemon/daemon.go`

File: daemon/runtime_unix.go [hunk 2]
```diff
@@ -47,92 +63,88 @@ func defaultV2ShimConfig(conf *config.Config, runtimePath string) *types.ShimCon
...
-	rt := cfg.GetRuntime(name)
-	if rt == nil {
-		if !config.IsPermissibleC8dRuntimeName(name) {
-			return nil, errdefs.InvalidParameter(errors.Errorf("unknown or invalid runtime name: %s", name))
-		}
-		return &types.Runtime{Shim: &types.ShimConfig{Binary: name}}, nil
+// A non-standard Base32 encoding which lacks vowels to avoid accidentally
+// spelling naughty words. Don't use this to encode any data which requires
+// compatibility with anything outside of the currently-running process.
+var base32Disemvoweled = base32.NewEncoding("0123456789BCDFGHJKLMNPQRSTVWXYZ-")
```
Comment (tianon) [inline daemon/runtime_unix.go:131]: :joy: :heart:

(Should this also remove "vowel looking" numbers like `0`, `1`, `3`, `4`?)

**Review by tianon — APPROVED** [review]
The implementation details look pretty hairy, but it does seem to be pretty comprehensive (including new tests :eyes:)

**Review by cpuguy83 — APPROVED** [review]

File: daemon/config/registry.go (hunk 1 above)
Comment (corhere) [reply daemon/config/registry.go]: Sadly, I had to move it back. On a positive note, reloaded registry configuration will now be picked up by Buildkit!

**Comment by corhere** [comment]
It was a nasty rebase, though on the bright side, I spotted and fixed some mistakes in the process and made some improvements.

File: api/server/router/build/build.go [hunk 3]
```diff
@@ -9,18 +9,16 @@ import (
 
 // buildRouter is a router to talk with the build controller
 type buildRouter struct {
-	backend  Backend
-	daemon   experimentalProvider
-	routes   []router.Route
-	features *map[string]bool
```
Comment (cpuguy83) [inline api/server/router/build/build.go:15]: 👀 🤦 
I totally missed that this was happening.

**Review by cpuguy83 — APPROVED** [review]

**Comment by thaJeztah** [comment]
discussing with @corhere - he wants to look at splitting out one fix from this PR to make it back-portable, so let's wait with merging for that.

**Comment by corhere** [comment]
@thaJeztah all done: the first two commits are now the backport-able fixes. I went through the diffs again to confirm that those are the only changes worth backporting.

**Review by Iv3-0 — COMMENTED** [review]
vendor/modules.txt

**Pull request merged**